cassiopeia/solver/SpectralNeighborJoiningSolver.py

```diff
@@ -0,0 +1,281 @@
+                a_comp.pop(min(i_count, j_count))
+                a_comp_flat = list(chain.from_iterable(a_comp))
+
+                # create RA matrix
+                RA_matrix = self.__similarity_map.values[
+                    np.ix_(a_subset, a_comp_flat)
+                    ]
+
+                # Calculate SVD2 #todo: remove alternatives
+                svd2_val = 0
+                s = scipy_svd(RA_matrix, compute_uv=False, check_finite=False)
+                if len(s) > 1:
+                    svd2_val = s[1]
+                else:
+                    svd2_val = s[0]
+                
+                    
+                lambda_matrix_arr[i_count, j_count] = lambda_matrix_arr[
+                    j_count, i_count] = svd2_val
+
+        np.fill_diagonal(lambda_matrix_arr, np.inf)
+
+        return lambda_matrix_arr
+
+    def update_dissimilarity_map(
+        self,
+        dissimilarity_map: pd.DataFrame, # lambda matrix
+        cherry: Tuple[str, str],
+        new_node: str,
+        ) -> pd.DataFrame:
+        """Updates the lambda matrix using the pair of subsets from find_cherry.
+
+        Args:
+            dissimilarity_map: lambda matrix to update
+            cherry1: One of the children to join.
+            cherry2: One of the children to join.
+            new_node: New node name to add to the dissimilarity map
+
+        Returns:
+            An updated lambda matrix.
+        """
+
+        lambda_indices_copy = self.lambda_indices.copy()
+
+        # get cherry nodes in index of lambda matrix
+        i, j = (
+            np.where(dissimilarity_map.index == cherry[0])[0][0],
+            np.where(dissimilarity_map.index == cherry[1])[0][0],
+        )
+
+        # modify names
+        node_names = dissimilarity_map.index.values
+        node_names = np.append(node_names, new_node)
+        node_names = np.delete(node_names, [i, j])
+
+        # modify indices
+        lambda_indices_copy.append([
+            *lambda_indices_copy[i], 
+            *lambda_indices_copy[j]
+            ])
+        lambda_indices_copy.pop(max(i, j))
+        lambda_indices_copy.pop(min(i, j))
+
+        if len(lambda_indices_copy) > 2:
+            # compute new lambda matrix
+            lambda_matrix_arr = self.compute_lambda(lambda_indices_copy)
+        else:
+            n = len(lambda_indices_copy)
+            lambda_matrix_arr = np.zeros((n, n))
+
+        # regenerate lambda matrix
+        lambda_matrix_df = pd.DataFrame(
+            lambda_matrix_arr,
+            index=node_names,
+            columns=node_names
+        )
+
+        # update lambda indices
+        self.lambda_indices = lambda_indices_copy
+
+        return lambda_matrix_df
+
+    def setup_root_finder(self, cassiopeia_tree: CassiopeiaTree) -> None:
+        """Defines the implicit rooting strategy for the NeighborJoiningSolver.
+
+        By default, the SpectralNeighborJoining algorithm returns an unrooted tree.
```

Comment:
Line length.

Also -- if this function is taken directly from `NeighborJoiningSolver`, make it clear that we are duplicating it here.